### ISSUE TITLE
Declare requirements in the tests to run\skip based on runtime checks

### ISSUE THREAD

**Issue by pallavit** [dotnet/corefx#3349]
As of now a no. of tests have requirements which make them applicable only in certain situations. For example, ServiceProcess tests can only be run in admin mode, several PLINQ tests need to be run on multi-threaded machines. (We do have an option of marking these tests OpenLoop, but that seems like a broader classification). Currently, the only way to assert these requirements is to update the constructor to check for the requirements, causing each individual test to fail. It would be good to have a means to skip\run tests based on a runtime check. #3134 has more details

**Comment by roncain** [comment]
Our requirements are not just for conditionally running based on runtime checks, but also on checks of the parameters passed to xunit when the run was launched.  We would use [ConditionalFact] to examine the way xunit was invoked to determine how to answer.

What is the current plan to feed information into the test run to allow something like [ConditionalFact] to make a choice depending on how the user asked to run the test?  Is there an existing way to pass parameters to xunit that tests can interrogate?  Do we settle on using environment variables?

For example, I want a developer to say things like:
* Run all tests except the ones that install certificates
* Run all tests except those requiring domain joined machines
* Run only the TCP tests

Is it reasonable to agree on an environment variable holding a semicolon separated list of options that express this, such as:
    set TestProperties=+TCP;-HTTP;-Certificates, or
    set TestProperties=All;-Domain

This becomes important when developers want to run subsets of tests.  Obviously each product might adopt its own conventions, but it seems a good idea to have a common way to do the same thing.

**Comment by stephentoub** [comment]
Thanks, Ron. You might want to open a separate issue for that; I don't want it to get lost when this closes.

Regarding your examples, isn't that what traits are meant for? Are there gaps there that need to be addressed?

@chrisboh can probably speak to our broader plans here, which I'm not entirely familiar with. I added these Conditional attributes as a way to unblock some of our current issues, and I expect they'll continue to be useful longer than just as a stop gap, but will see :)

**Comment by roncain** [comment]
Thanks @stephentoub .  Yes, I would like to hear thoughts from @Chrisboh on the all-up plan for test filtering/skipping.  I think [ConditionalFact] and [ConditionalTheory] will be valuable for multiple uses, including some that overlap they way traits might be used.  WCF will almost certainly use [ConditionalFact] to auto-skip when communicating with services that are not domain joined.  This seems far friendlier than telling developers "be sure to exclude tests with trait XYZ if you are not using a domain-joined service."
